packages/service-library/src/servicelib/utils.py

```diff
@@ -175,3 +176,115 @@ def unused_port() -> int:
     with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
         s.bind(("127.0.0.1", 0))
         return cast(int, s.getsockname()[1])
+
+
+T = TypeVar("T")
+
+
+async def limited_as_completed(
+    awaitables: Iterable[Awaitable[T]] | AsyncIterable[Awaitable[T]], *, limit: int = 1
+) -> AsyncGenerator[asyncio.Future[T], None]:
```

Comment:
Very cool function! And very nice PR!

One thing I can't help ask myself is the following: This function is probably called by a coroutine which is itself running in the event loop, so this function already might be called concurrently and then the size of the event loop can also explode. Is there a way to limit the total number of tasks/futures in the event loop? Like the total number of threads when doing threading.

Reply:
well this sounds like a very difficult task ;)
I guess it all depends on how you use it right? if you do simple things, you could as well put no limits at all.
if you nest a lot of calls and end up moving 1000s of files, then it can become a hurdle.... not sure there is an easy automated way to limit it.